Feature/cookies session

plwai opened the pull request:
- [x] Cookies

- [x] Session Structure

- [x] Session middleware

resolve #42

plwai committed:
feat: impl set cookies

Implement set cookies using cookies trait as there are a lot of options within cookie properties.

plwai committed:
feat: impl cookie parser

Implement cookie parser middleware. Enable context cookie getter. Fix
context dynamic data get and get_mut return wrong type issue.

plwai commented:
Blocked until #47 completed.

pickfire commented on src/context.rs:
```diff
@@ -269,6 +271,16 @@ impl Context {
 
+    pub fn cookie(&self, name: &str) -> Option<&Cookie> {
+        if let Some(cookie_data) = self.get::<CookieParserData>() {
+            if let Some(ref cookie) = cookie_data.cookie_jar().get(name) {
+                return Some(*cookie);
+            }
```
It it possible to just use:
```rust
            return cookie_data.cookie_jar().get(name)
```

pickfire commented on src/middleware/cookie_parser.rs:
```diff
@@ -0,0 +1,68 @@
+
+                cookie_str
+                    .split("; ")
+                    .map(|x| x.trim().splitn(2, '=').collect::<Vec<&str>>())
```
Shouldn't it collect as tuple? This does `Vec` allocation in every loop inside `map`.

plwai replied:
I was thinking on if there is extra '=' being sent in the http header. Anyway, the whole `cookies session` will be revamped as the core flow is being changed.

pickfire commented on src/router/handler.rs:
```diff
@@ -26,7 +26,26 @@ where
+
+        if let Some(cookies) = response.cookies() {
+            if let Some(response_headers) = res.headers_mut() {
+                cookies.iter().for_each(move |cookie| {
```
How many cookies are there to iterate? I thought there should be only one cookie?

plwai replied:
`set-cookie` header can be multiple.

pickfire commented on src/router/response.rs:
```diff
@@ -1,14 +1,16 @@
     body: Body,
     status: StatusCode,
-    headers: Option<Vec<(header::HeaderName, &'static str)>>,
+    headers: Option<Vec<(header::HeaderName, String)>>,
```
Why not use `Cow<&'static str>` here to allow both `&str` and `String` with just little computation cost? If we want it to be smaller, we can use https://crates.io/crates/beef.

pickfire commented on src/router/response.rs:
```diff
@@ -32,11 +35,15 @@ impl Response {
     }
 
-    pub fn headers(&self) -> &Option<Vec<(header::HeaderName, &'static str)>> {
+    pub fn headers(&self) -> &Option<Vec<(header::HeaderName, String)>> {
```
Same for this, it can also be `Cow` or `AsRef<str>` to make use of both.

plwai replied:
Failed to do this for multiple headers as `Cow` does not implement `Copy`. Conversion from `Into<Cow<'static, str>>` is not possible.

pickfire commented on src/router/response.rs:
```diff
@@ -54,36 +61,69 @@ impl Response {
+        self
+    }
+
+    pub fn set_cookies(mut self, mut cookies: Vec<Cookie<'static>>) -> Self {
```
Shouldn't this take `&[Cookie<'static>]`?

pickfire commented on src/router/response.rs:
```diff
@@ -54,36 +61,69 @@ impl Response {
+        self
+    }
+
+    pub fn set_headers(mut self, headers: Vec<(header::HeaderName, &str)>) -> Self {
```
Shouldn't this take `&[(header::HeaderName, &str)]`?

plwai replied:
It set to `Vec` due to lifetime issue i guess. This is from @jk-gan #34.

pickfire replied:
Oh

jk-gan commented on Cargo.toml:
```diff
@@ -33,6 +33,7 @@ serde_json = '1.0.44'
 url = '2.1.0'
 async-std = '1.4.0'
 async-trait = '0.1.22'
+cookie = "0.13.3"
```
```suggestion
// testing
cookie = "0.13.3"
```

plwai committed:
feat: revamp cookies parser flow

Revamp the flow to fit current routing handler. Cookie flow and
convertion are improved in this commit too.

plwai committed:
feat: implement memory session

plwai committed:
refactor: apply clippy
